internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go

```diff
@@ -46,67 +55,189 @@ func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed boo
 		return nil, errors.New(err, "failed to unpack configuration")
 	}
 	return &contextProviderK8sSecrets{
-		logger: logger,
-		config: &cfg,
+		logger:       logger,
+		config:       &cfg,
+		secretsCache: make(map[string]*secretsData),
 	}, nil
 }
 
 func (p *contextProviderK8sSecrets) Fetch(key string) (string, bool) {
-	// key = "kubernetes_secrets.somenamespace.somesecret.value"
+	return p.getFromCache(key)
+}
+
+// Run initializes the k8s secrets context provider.
+func (p *contextProviderK8sSecrets) Run(ctx context.Context, comm corecomp.ContextProviderComm) error {
+	client, err := getK8sClientFunc(p.config.KubeConfig, p.config.KubeClientOptions)
+	if err != nil {
+		p.logger.Debugf("Kubernetes_secrets provider skipped, unable to connect: %s", err)
+		return nil
+	}
 	p.clientMx.Lock()
-	client := p.client
+	p.client = client
 	p.clientMx.Unlock()
-	if client == nil {
-		return "", false
+	go p.updateSecrets(ctx)
+
+	<-comm.Done()
+
+	p.clientMx.Lock()
+	p.client = nil
+	p.clientMx.Unlock()
+	return comm.Err()
+}
+
+func getK8sClient(kubeconfig string, opt kubernetes.KubeClientOptions) (k8sclient.Interface, error) {
+	return kubernetes.GetKubernetesClient(kubeconfig, opt)
+}
+
+// Update the secrets in the cache every RefreshInterval
+func (p *contextProviderK8sSecrets) updateSecrets(ctx context.Context) {
+	timer := time.NewTimer(p.config.RefreshInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
+			p.updateCache()
+			timer.Reset(p.config.RefreshInterval)
+		}
+	}
+}
+
+func (p *contextProviderK8sSecrets) updateCache() {
+	// deleting entries does not free the memory, so we need to create a new map
+	// to place the secrets we want to keep
+	cacheTmp := make(map[string]*secretsData)
+
+	p.secretsCacheMx.RLock()
+	for name, data := range p.secretsCache {
+		diff := time.Since(data.lastAccess)
+		if diff < p.config.TTLDelete {
+			value, ok := p.fetchSecretWithTimeout(name)
+			if ok {
+				newData := &secretsData{
+					value:      value,
+					lastAccess: data.lastAccess,
+				}
+				cacheTmp[name] = newData
+			}
+
+		}
+	}
+	p.secretsCacheMx.RUnlock()
+
+	p.secretsCacheMx.Lock()
+	p.secretsCache = cacheTmp
+	p.secretsCacheMx.Unlock()
+}
+
+func (p *contextProviderK8sSecrets) getFromCache(key string) (string, bool) {
+	p.secretsCacheMx.RLock()
+	_, ok := p.secretsCache[key]
+	p.secretsCacheMx.RUnlock()
+
+	// if value is still not present in cache, it is possible we haven't tried to fetch it yet
+	if !ok {
+		data, ok := p.addToCache(key)
+		// if it was not possible to fetch the secret, return
+		if !ok {
+			return data.value, ok
+		}
 	}
+
+	var pass string
+	p.secretsCacheMx.Lock()
+	data, ok := p.secretsCache[key]
+	data.lastAccess = time.Now()
```

Comment:
`data` is not a pointer so you are getting a copy of it, so the update to `lastAccess` does not change the `lastAccess` that was originally stored in the map.

You are missing test coverage for two consecutive `Fetch` calls of the same value to ensure that `lastAccess` updates in the expected way.

Here's a sample modification to your test showing the problem:

```diff
diff --git a/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets_test.go b/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets_test.go
index 3fea2dc455..a7348c40a9 100644
--- a/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets_test.go
+++ b/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets_test.go
@@ -212,12 +212,33 @@ func Test_K8sSecretsProvider_Check_TTL(t *testing.T) {
        val, found := fp.Fetch(key)
        assert.True(t, found)
        assert.Equal(t, val, pass)
+
        fp.secretsCacheMx.RLock()
        assert.Equal(t, len(fp.secretsCache), 1)
-       assert.NotNil(t, fp.secretsCache[key])
-       assert.NotZero(t, fp.secretsCache[key].lastAccess)
+       cachedVal1 := fp.secretsCache[key]
        fp.secretsCacheMx.RUnlock()

+       assert.NotNil(t, cachedVal1)
+       assert.NotZero(t, cachedVal1.lastAccess)
+
+       val, found = fp.Fetch(key)
+       assert.True(t, found)
+       assert.Equal(t, val, pass)
+
+       // Arbitrary delay to ensure that the value returned by time.Now() used when setting TTL is not
+       // the same between Fetch() calls.
+       time.Sleep(time.Millisecond)
+
+       fp.secretsCacheMx.RLock()
+       assert.Equal(t, len(fp.secretsCache), 1)
+       cachedVal2 := fp.secretsCache[key]
+       fp.secretsCacheMx.RUnlock()
+
+       assert.NotNil(t, cachedVal2)
+       t.Logf("Fetch 1 LastAccess: %s", cachedVal1.lastAccess)
+       t.Logf("Fetch 2 LastAccess: %s", cachedVal2.lastAccess)
+       assert.NotEqual(t, cachedVal1.lastAccess, cachedVal2.lastAccess)
+
        // Update the secret and check after TTL time, the secret value is correct
        newPass := "new-pass"
        secret = &v1.Secret{

```

```sh
❯ go test ./internal/pkg/composable/providers/kubernetessecrets/... -run Test_K8sSecretsProvider_Check_TTL

# github.com/elastic/elastic-agent/internal/pkg/composable/providers/kubernetessecrets.test
ld: warning: ignoring duplicate libraries: '-lproc'
--- FAIL: Test_K8sSecretsProvider_Check_TTL (2.22s)
    kubernetes_secrets_test.go:238: Fetch 1 LastAccess: 2023-12-05 16:35:13.749574 -0500 EST m=+0.016378918
    kubernetes_secrets_test.go:239: Fetch 2 LastAccess: 2023-12-05 16:35:13.749574 -0500 EST m=+0.016378918
    kubernetes_secrets_test.go:240:
                Error Trace:    /Users/cmackenzie/go/src/github.com/elastic/elastic-agent/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets_test.go:240
                Error:          Should not be: time.Date(2023, time.December, 5, 16, 35, 13, 749574000, time.Local)
                Test:           Test_K8sSecretsProvider_Check_TTL
FAIL
FAIL    github.com/elastic/elastic-agent/internal/pkg/composable/providers/kubernetessecrets    2.754s
FAIL
```

Reply:
Thanks @cmacknz for such a complete example.

I checked this:

> data is not a pointer so you are getting a copy of it, so the update to lastAccess does not change the lastAccess that was originally stored in the map.

And I ran the test. I added a println on the code right where the last access changes:

```
=== RUN   Test_K8sSecretsProvider_Check_TTL
2023-12-06 08:26:10.267182544 +0100 CET m=+0.016935938
2023-12-06 08:26:10.267266524 +0100 CET m=+0.017019919
    kubernetes_secrets_test.go:239: Fetch 1 LastAccess: 2023-12-06 08:26:10.267266524 +0100 CET m=+0.017019919
    kubernetes_secrets_test.go:240: Fetch 2 LastAccess: 2023-12-06 08:26:10.267266524 +0100 CET m=+0.017019919
    kubernetes_secrets_test.go:241: 
        	Error Trace:	/home/c/Documents/Git-Code/Elastic-Forks/elastic-agent/internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets_test.go:241
        	Error:      	Should not be: time.Date(2023, time.December, 6, 8, 26, 10, 267266524, time.Local)
        	Test:       	Test_K8sSecretsProvider_Check_TTL
--- FAIL: Test_K8sSecretsProvider_Check_TTL (0.01s)
```

So, as you can see, the last value got updated. But since `cachedVal1` and `cachedVal2` are the same, this assertion will always fail `assert.NotEqual(t, cachedVal1.lastAccess, cachedVal2.lastAccess)`. Notice that the map contains the exact values, not copies:

```
type contextProviderK8sSecrets struct {
        ...
	secretsCache   map[string]*secretsData <---------------------- pointer here
}
```

Reply:
Yes you are correct, my example proves myself wrong I just needed to read it at the start of my day instead of the end :)